design-proposals/vsock/vsock.md

```diff
@@ -0,0 +1,92 @@
+
+### Design
+
+The feature can be added to KubeVirt in two phases.
+
+*   Phase 1: node-local bits to wire in vsock, only allowing privileged node-local services to communicate with the Guest VMs.
+*   Phase 2: Add websocket proxying to virt-handler, to allow accessing VSOCK ports via REST calls and  protecting it via k8s RBAC.
+
+### Phase 1: node-local wiring of vsock
+
+*   The `/dev/vhost-vsock` device will be exposed via virt-handler under the `vsock` feature gate
+*   Each virt-launcher with `vsock` enabled will request the device
+*   Virt-controller will hand out a CID to virt-launcher and add it to the VMI status. 
+*   virt-launcher/libvirt/qemu take the CID and configure the vsock device via `/dev/vhost-vsock`
+*   From this point on `/dev/vhost-vsock` does not have any function anymore
+*   To simplify the migration story, the CID will be unique cluster wide. Therefore the CID will be retained after migration. (With 100 000 k8s nodes we would still have enough CIDs to run more than 20 000 VMs per node). The CID has a type of `uint32`, which is large enough to satisfy a cluster-wide unique allocation)
+
+![vsock flow](vsock.png)
+
+#### API changes
+
+A `autoattachVsock` flag will be added to the VMI spec at `spec.domain.devices` which defaults to `false` to stay backward compatible. Also a `VsockCID` field will be added to the VMI status as `status.vsockCID` for CID allocation and lookup.
+
+### Phase 2: websocket access via the apiserver
+
+This follows exactly the same architecture as vnc/console. `GET /virtualmachineinstances/&lt;name>/vsock?port=123` will be added to the REST-API and it will return a websocket connection to the guest, if something is listening on the guest. Then, for example, node-local or control-plane services can establish a gRPC connection on top of the websocket connection to the guest service and invoke RPC calls.
+
+By default none of our aggregated cluster-roles will get RBAC access to this endpoint.
+
+Creating a websocket connection via this REST-API means that all traffic goes
+through the kubernetes apiserver. Therefore the same scale considerations apply
+like for `exec`, `port-forward`, `vnc` and so forth.
+
+### Security
+
+Note: The `dev/vhost-vsock` device is **NOT NEEDED** to connect or bind to a `vsock` port. This means that it is slightly more challenging to secure it.
+
+The following measure will therefore be put in place:
+
+*   The whole `vsock` features will live behind a feature gate
+*   By default the first 1024 ports of a vsock device are privileged. Services trying  to connect to those require `CAP_NET_BIND_SERVICE`.
+*   We aim to block the `AF_VSOCK` `socket` syscall in containerd (https://github.com/containerd/containerd/issues/7442), however if that will happen is not sure. It is therefore right now the responsibility of the vendor to apply the required seccomp change to all unprivileged ports, or to verify that the `CAP_NET_BIND_SERVICE` privilege is handed out carefully enough
```

Comment:
Why did you choose containerd? I think containers/common would also benefit and might even reach a broader spectrum of users. (Kubervirt itself doesn't use containerd for some time...)

Reply:
Yes, but we use containerd. This is an independent effort of this proposal to secure it for our purpose. I think it makes a lot of sense to also push this to other popular CRIs, but it is out of scope of this proposal :)

Reply:
I see. I read the proposal as a community - meaning we == community, and in that sense, it would make more sense to aim for changes in crio(therefore containers/common). I can pick this up.

Reply:
Ah, I see the confusion now regarding to the usage of `we`. :+1: